File: JavaScript/298-Binary Tree Longest Consecutive Sequence.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number}
 */
var longestConsecutive = function(root) {
    res = 0;
    
    let dfs = (node) =>{
        if (!node) return 0;
        
        let length = 1
        
        if (node.left){
            if (node.val + 1 === node.left.val){
                length = dfs(node.left) + 1;
            }else{
                dfs(node.left);
            }
        }
        
        if (node.right){
            if (node.val + 1 === node.right.val){
                length = Math.max(length, dfs(node.right)+1);
            }else{
                dfs(node.right);
            }
        }
        
        res = Math.max(res, length);
        return length;
    }
    dfs(root);
    return res;
};

File: C++/298-Binary Tree Longest Consecutive Sequence.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    int res = 1;
    void dfs(TreeNode* node, int parentVal, int length) {
        int val = node->val;
        if (val == parentVal+1) {
            length++;
            res = max(res, length);
        } else {
            length = 1;
        }

        if (node->left != nullptr) {
            dfs(node->left, val, length);
        }
        if (node->right != nullptr) {
            dfs(node->right, val, length);
        }

    }
    int longestConsecutive(TreeNode* root) {
        dfs(root, 3e4+1, 0);

        return res;
    }
};